### ISSUE TITLE
Autocorrect Style/HashTransformValues is broken

### ISSUE THREAD

**Issue by mdemare** [rubocop-hq/rubocop#7767]
Firstly, it transforms some usages of `each_with_object` where the receiver is an Array.
Secondly, looking up a missing value in the resulting hash will return a `Proc` object, instead of `nil`.

This is _not_ a safe transformation.

**Comment by koic** [comment]
`Style/HashTransformValues` is marked as unsafe.
https://github.com/rubocop-hq/rubocop/blob/v0.80.1/config/default.yml#L2824

Using `rubocop --safe` and `rubocop --safe-auto-correct` command-line options run to only safe cops. So users can omit unsafe cops using this option.

By the way, could you show me the reproduction example code of the issue?

**Comment by lucascaton** [comment]
Hi @koic! 👋

I have a reproduction example code:

```ruby
[[1, 2], [3, 4]].map { |k, v| [k, [v, nil]] }.to_h
# => {1=>[2, nil], 3=>[4, nil]}
```

When I run `$ rubocop`, this is the output:

```sh
code.rb:3:1: C: Style/HashTransformValues: Prefer transform_values over map {...}.to_h.
[[1, 2], [3, 4]].map { |k, v| [k, [v, nil]] }.to_h
```

And running `$ rubocop -a`, it auto-fixes it to:

```diff
-[[1, 2], [3, 4]].map { |k, v| [k, [v, nil]] }.to_h
+[[1, 2], [3, 4]].transform_values { |v| [v, nil] }
```

However, the produced (above) code isn't valid:

```ruby
[[1, 2], [3, 4]].transform_values { |v| [v, nil] }
# NoMethodError: undefined method `transform_values' for [[1, 2], [3, 4]]:Array
```

**Comment by tejasbubane** [comment]
@koic I think this cop should skip array literals?

**Comment by koic** [comment]
Surely Array instance does not have the `transform_values` method, so I agree to skip for array literals.

**Comment by koic** [comment]
Also the `transform_keys` method (`Style/HashTransformKeys` cop) is the same.

**Comment by fsoikin** [comment]
This is not limited to literals though. In our codebase I found hit rate of this rule (and other similar hash-related rules) to be about 60% overall. For example, this is also considered a hit:

```
{ x: 1,  y: 5 }.sort_by(&:first).to_h { |k, v| [k, v * 2] }
```

Even though `sort_by` returns an `Array`, not a `Hash`.

Just in general, this check reacts to the code pattern, but doesn't always make sure that the receiver object is actually a hash.
